apps/remote_control/lib/lexical/remote_control/search/store/backends/cub/state.ex

```diff
@@ -0,0 +1,242 @@
+defmodule Lexical.RemoteControl.Search.Store.Backends.Cub.State do
+  @moduledoc """
+  A cubdb schema and functions for data manipulation
+
+  CubDB is an... interesting database. It relies on erlang term ordering, which is documented below.
+  Furthermore, the positions of the records are significant in picking the ranges of things that we
+  must iterate over in order to limit the amount of stuff each `select` call sees.
```

Comment:
Suggested rewrite:

CubDB is a KV store that stores data sorted using Erlang term ordering:

    number < atom < reference < fun < port < pid < tuple < map < nil < list < binary

The main query method is `select` using a min and max key. We use records as keys, which allow us to limit the data we scan by specifying exact values at the "front" of the record and relying on term order to construct wildcard queries at the "back" of the record. 

For instance, if we've stored the following keys:

    {:subject_key, :module, :definition, {"Foo"}}
    {:subject_key, :module, :definition, {"Bar"}}
    {:subject_key, :module, :reference, {"Foo"}}
    
We could select all module definitions using the min and max keys:

    iex> CubDB.select(db,
    ...>   min_key: {:subject_key, :module, :definition, {}},
    ...>   max_key: {:subject_key, :module, :definition, {"", ""}}
    ...> ) |> Enum.map(&elem(&1, 0))
    [{:subject_key, :module, :definition, {"Bar"}},
     {:subject_key, :module, :definition, {"Foo"}}]

(Note that we've wrapped the subject in a tuple. Because strings compare greater than all other terms, wrapping them in a tuple allows us to create max keys greater than all subjects by using a larger tuple, which always compares greater than a smaller tuple regardless of the contained terms.)

Reply:
BTW this is written assuming we need to do the "wrap binaries in a tuple" trick. Can remove/adjust that part if we don't actually need to do that.

Reply:
We still might need to do that. This is a draft to get the ideas out there.